Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+   */
+  template<class ...ARGS>
+  void initialize_or_refresh(ARGS... args){
+    if( initialized() ){
+      refresh(args...);
+    } else {
+      initialize(args...);
+    }
+  }
+
+  /*! \brief Initialization status of the NdFlattener.
+   * \returns true if the NdFlattener has been initialized
+   */
+  bool initialized(){
+    return nNodes>0;
+  }
+
+protected:
+  /*! \brief Allocate the indices array after \a nNodes has been determined.
+   */
+  void allocate() {
+    indices.reserve(nNodes+1);
+    indices[0] = 0;
+    Base::allocate();
+  }
+
+  /*! \brief Set \a iNode to 0 in all layers.
+   */
+  void reset_iNode(){
+    iNode = 0;
+    Base::reset_iNode();
+  }
+
+  /*=== Construct from 'recursive function' ===*/
+public:
+  /*! \brief Initialize from a 'recursive function'.
+   *
+   * The function should return a pair. Its first entry is the number of children. 
+   * Its second entry is a function with the same meaning, recursively 
+   * one layer down.
+   * \param f - the 'recursive function'
+   */
+  template<class f_type>
+  void initialize(f_type f) {
+    count_f(f);
+    allocate();
+    set_f(f, false);
+  }
+
+  /*! \brief Refresh the data according to the 'recursive function'
+   *
+   * The NdFlattener must have been constructed with a 'recursive function'.
+   * Now refresh the values with another 'recursive function'. The subarray lengths 
+   * resulting from both 'recursive functions' must coincide, as the indices arrays
+   * are not changed.
+   * 
+   * \param f - the 'recursive function'
+   * \tparam f_type - to allow for any type of the 'recursive function'
+   */
+  template<class f_type>
+  void refresh(f_type f){
+    reset_iNode();
+    set_f(f, true);
+  }
+
+protected:
+  /*! \brief Determine the space required for reading the 'recursive function'.
+   *
+   * \param f - the 'recursive function'
+   */
+  template<class f_type>
+  void count_f(f_type f) {
+    Index nChild = f.first;
+    for(Index iChild=0; iChild<nChild; iChild++){
+      nNodes++;
+      Base::count_f(f.second(iChild));
+    }
+  }  
+
+  /*! \brief Read the 'recursive function' into the allocated arrays.
+   *
+   * \param f - the 'recursive function'
+   * \param refresh - if true, the object is already initialized and only the data
+   *   in layer 1 have to be overwritten
+   * \tparam f_type - to allow for any type of the 'recursive function'
+   */
+  template<class f_type>
+  void set_f(f_type f, bool refresh) {
+    Index nChild = f.first;
+    for(Index iChild=0; iChild<nChild; iChild++){
+      Base::set_f(f.second(iChild), refresh);
+      if(!refresh){
+        indices[iNode+1] = indices[iNode] + f.second(iChild).first;
+      } else {
+        if( indices[iNode+1] != indices[iNode] + f.second(iChild).first ){
+          SU2_MPI::Error("NdFlattener: Structure has changed, cannot refresh.", CURRENT_FUNCTION);
+        }
+      }
+      iNode++;
```

Comment:
iNode = iChild ?

Reply:
No, for the same reason as above. `iNode` is set to 0 only when initializing or refreshing the NdFlattener. Whenever in this depth-first search you descent to the first child node, the `iChild` in some layer is set to 0.